### ISSUE TITLE
export-client can't notify export-distro

### ISSUE THREAD

**Issue by AhaEdgar** [edgexfoundry/edgex-go#755]
After the export-client service is started, it will get Services Info from consul. However, at this time, export-distro service or rules-engine service has not started. Hence, export-client service couldn't get export-distro service info from consul. Besides, if I want to register a new export client via POST http://localhost:48071/api/v1/registration at this time, the logs of export-client will print ERROR PUT http:// : //api/v1/notify/registrations.

Two problem:

1. why ipaddress and port are null ?
The 'ClientMonitor' in 'cmd/export-client/res/docker/configuration.toml' is 15000. It means that export-client service will wait 15000s, then it will get latest Services Info from consul. Hence, when the export-client service first started, it doesn't get the export-distro service info. Only after waiting 15000s, the export-client will get latest Service Info from consul. So I modify the 'ClientMonitor' to reduce the waiting time.
in 'cmd/export-client/res/docker/configuration.toml'
```
[Service]
BootTimeout = 30000
//ClientMonitor = 15000
ClientMonitor = 300 
CheckInterval = '10s'
Host = 'edgex-export-client'
Port = 48071
```

2. error url - " http:// : //api/v1/notify/registrations"

```
[root@localhost compose-files]# docker logs -t -f --tail 20 27ea8c888ab5
2018-11-07T12:54:37.663270732Z INFO: 2018/11/07 12:54:37 Service dependencies resolved...
2018-11-07T12:54:37.663347214Z INFO: 2018/11/07 12:54:37 Starting edgex-export-client 1.0.0 
2018-11-07T12:54:37.663361198Z INFO: 2018/11/07 12:54:37 This is the Export Client Microservice
2018-11-07T12:54:37.663372615Z INFO: 2018/11/07 12:54:37 Service started in: 27.241579ms
2018-11-07T12:54:37.663381884Z INFO: 2018/11/07 12:54:37 Listening on port: 48071
2018-11-07T12:57:35.504670246Z ERROR: 2018/11/07 12:57:35 error from distro: Put http://:0//api/v1/notify/registrations: dial tcp :0: getsockopt: connection refused
```

There are two "/" in front of the "api " , so export-client couldn't request "PUT http:// : //api/v1/notify/registrations". Hence, I remove the "/" in params

in internal/export/client/init.go

```
	// Create export-distro client
	params := types.EndpointParams{
		ServiceKey:  internal.ExportDistroServiceKey,
		//Path:        "/",
		Path:        "",
		UseRegistry: useConsul,
		Url:         Configuration.Clients["Distro"].Url(),
		Interval:    Configuration.Service.ClientMonitor,
```

**Comment by tsconn23** [comment]
@AhaEdgar How are you starting the services? Using the recommended `make run` the condition you describe never happens.

```
INFO: 2018/11/07 10:02:23 Starting edgex-export-client 1.0.0 
INFO: 2018/11/07 10:02:23 This is the Export Client Microservice
INFO: 2018/11/07 10:02:23 Service started in: 6.545065ms
INFO: 2018/11/07 10:02:23 Listening on port: 48071
INFO: 2018/11/07 10:02:24 Service dependencies resolved...
INFO: 2018/11/07 10:02:24 Starting edgex-export-distro 1.0.0 
INFO: 2018/11/07 10:02:24 This is the Export Distro Microservice
INFO: 2018/11/07 10:02:24 Starting Export Distro :48070
INFO: 2018/11/07 10:02:24 Connecting to incoming 0MQ at: tcp://localhost:5563
INFO: 2018/11/07 10:02:24 Connected to inbound 0MQ
INFO: 2018/11/07 10:02:24 Starting registration loop
```

In addition, this never happens when bringing the services via docker compose. If you are starting the services manually, engendering a delay between client and distro, then you are welcome to change the  ClientMonitor value to suit your needs. You would also be welcome to create a separate issue and submit a PR providing a retry mechanism at the point of perceived failure. However we will not be changing the config value in the repo which provides a default for everyone to use.

The second point related to the Path property is valid, but doesn't appear to cause a runtime error meaning the double slash is being escaped. If this were causing an error, you would see it in the log output above. Specifically the following line wouldn't show up and there would be an error instead.

`INFO: 2018/11/07 10:02:24 Starting registration loop`

You are welcome to create another issue containing that specific item and a PR with that specific change.

**Comment by AhaEdgar** [comment]
@tsconn23 I am starting the services via run command `make run_docker `. 

```
INFO: 2018/11/07 10:02:23 Starting edgex-export-client 1.0.0 
INFO: 2018/11/07 10:02:23 This is the Export Client Microservice
INFO: 2018/11/07 10:02:23 Service started in: 6.545065ms
INFO: 2018/11/07 10:02:23 Listening on port: 48071
INFO: 2018/11/07 10:02:24 Service dependencies resolved...
INFO: 2018/11/07 10:02:24 Starting edgex-export-distro 1.0.0 
INFO: 2018/11/07 10:02:24 This is the Export Distro Microservice
INFO: 2018/11/07 10:02:24 Starting Export Distro :48070
INFO: 2018/11/07 10:02:24 Connecting to incoming 0MQ at: tcp://localhost:5563
INFO: 2018/11/07 10:02:24 Connected to inbound 0MQ
INFO: 2018/11/07 10:02:24 Starting registration loop
```
I think you get these logs via run command `make run`. It means you don't use consul, so your code wouldn't run this method (https://github.com/edgexfoundry/edgex-go/blob/76cda22cd5e8b8b7973e0b98e3a75198b26a434a/internal/export/client/init.go#L155). So, you wouldn't see the double slash problem.

If you `docker-compose down` all services, and then run `make run_docker` starting the services. When all services up, I use the  postman  to add a export-client (POST http://localhost:48071/api/v1/registration). Last, run 'docker logs -t -f --tail 20 container_id'，you can see the double slash problem. I have test the latest edgex-go(master) in this way, the error still happens.

```
[root@localhost compose-files]# docker logs -t -f --tail 20 27ea8c888ab5
2018-11-07T12:54:37.663270732Z INFO: 2018/11/07 12:54:37 Service dependencies resolved...
2018-11-07T12:54:37.663347214Z INFO: 2018/11/07 12:54:37 Starting edgex-export-client 1.0.0 
2018-11-07T12:54:37.663361198Z INFO: 2018/11/07 12:54:37 This is the Export Client Microservice
2018-11-07T12:54:37.663372615Z INFO: 2018/11/07 12:54:37 Service started in: 27.241579ms
2018-11-07T12:54:37.663381884Z INFO: 2018/11/07 12:54:37 Listening on port: 48071
2018-11-07T12:57:35.504670246Z ERROR: 2018/11/07 12:57:35 error from distro: Put http://:0//api/v1/notify/registrations: dial tcp :0: getsockopt: connection refused
```

**Comment by BenjaminITS** [comment]
Same issue for me, using docker-compose with the last version (0.7.0).
I tested with a fresh start and launching EdgeX in this order :
```
docker-compose up -d volume
docker-compose up -d consul
docker-compose up -d config-seed
docker-compose up -d mongo
docker-compose up -d logging
docker-compose up -d notifications
docker-compose up -d metadata
docker-compose up -d data
docker-compose up -d command
docker-compose up -d scheduler
docker-compose up -d export-client
docker-compose up -d export-distro
docker-compose up -d rulesengine
docker-compose up -d device-modbus
```

After adding my registration with POST http://localhost:48071/api/v1/registration, then I have the error in export-client :
```
INFO: 2018/11/22 09:45:17 Service dependencies resolved...
INFO: 2018/11/22 09:45:17 Starting edgex-export-client 0.7.0
INFO: 2018/11/22 09:45:17 This is the Export Client Microservice
INFO: 2018/11/22 09:45:17 Service started in: 589.100377ms
INFO: 2018/11/22 09:45:17 Listening on port: 48071
ERROR: 2018/11/22 09:45:49 error from distro: Put http://:0//api/v1/notify/registrations: dial tcp :0: getsockopt: connection refused
```

And this is the export distro logs :
```
INFO: 2018/11/22 09:45:21 Service dependencies resolved...
INFO: 2018/11/22 09:45:21 Starting edgex-export-distro 0.7.0
INFO: 2018/11/22 09:45:21 This is the Export Distro Microservice
INFO: 2018/11/22 09:45:21 Starting Export Distro :48070
INFO: 2018/11/22 09:45:21 Connecting to incoming 0MQ at: tcp://edgex-core-data:5563
INFO: 2018/11/22 09:45:21 Connected to inbound 0MQ
INFO: 2018/11/22 09:45:21 Starting registration loop
```

**Comment by tsconn23** [comment]
Closing due to inactivity

**Comment by michaelkargl** [comment]
Also running into this exact problem while following the Walkthrough

* https://fuji-docs.edgexfoundry.org/Ch-WalkthroughExporting.

Has a solution been found for this?

**Comment by cloudxxx8** [comment]
@michaelkargl export service has been deprecated.
could you please try to use app service?
https://docs.edgexfoundry.org/1.2/microservices/application/AppServiceConfigurable/
